Add generic container

S-Dafarra opened the pull request:
The basic idea is to have a class containing a span and a lambda to resize the original container. It is a first rough implementation and needs to be properly tested.

Contributions are welcome.

GiulioRomualdi commented:
Thanks, @S-Dafarra for the contribution ❤️

S-Dafarra commented:
I am finishing creating some easy constructors..usual SFINAE problems :sweat_smile:

S-Dafarra commented:
I have added also a simple test with ``yarp`` vectors. With the function ``makeGenericContainer``, it is sufficient to pass the container by reference to build a ``GenericContainer``, which is templated only on the value type..

S-Dafarra commented:
Tests are passing, I think it is ready to review. I still needs some documentation and additional test cases. Any suggestion?

GiulioRomualdi commented on src/GenericContainer/tests/GenericContainerTest.cpp:
```diff
@@ -0,0 +1,79 @@
+        iDynTree::VectorFixSize<3> fixedVector;
+        GenericContainer container2 = makeGenericContainer(fixedVector, GenericContainerMode::Resizable);
+        REQUIRE_FALSE(container.resize(2));
+    }
```
Correct me if I'm mistaken. `makeGenericContainer` generates a container that is not resizable. If you want to resize it `GenericContainerMode::Resizable` has to be added.  However, if you pass a vector that it is `FixedSize` the `resize` method it will return false

S-Dafarra replied:
Yes, you are right.

GiulioRomualdi commented on src/GenericContainer/tests/GenericContainerTestPlusYarp.cpp:
```diff
@@ -0,0 +1,81 @@
+
+        containerToBeCopied = container;
+
+        for (long i = 0; i < container.size(); ++i)
```
Why `long`?

S-Dafarra replied:
Apparently that's the ``index_type`` given by ``Span`` with ``double`` vectors. I discovered it with the static analyzer.

GiulioRomualdi commented on src/YarpUtilities/include/BipedalLocomotionControllers/YarpUtilities/Helper.h:
```diff
@@ -20,6 +20,8 @@
 
 #include <iDynTree/Core/VectorDynSize.h>
 
+#include <BipedalLocomotionControllers/TemplateHelpers.h>
```
Why not `#include <BipedalLocomotionControllers/GenericContainer/TemplateHelpers.h>`?

According to `add_bipedal_component` `cmake` function the files are installed in 
https://github.com/dic-iit/bipedal-locomotion-controllers/blob/e055599aa3cbe6bc3f4976a53c358694a5c0970a/cmake/AddBipedalLocomotionControllersComponent.cmake#L87

S-Dafarra replied:
I used only the version in the build folder actually. I did not notice. Is it ok for you if we include there a file of another component?

S-Dafarra replied:
Done in https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29/commits/2fd26da1dcaac5814f255c8e7d2ba2cea0dcae15

GiulioRomualdi commented on src/GenericContainer/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+# set target name
+add_bipedal_component(
+    NAME                   GenericContainer
+    PUBLIC_HEADERS         include/BipedalLocomotionControllers/GenericContainer.h include/BipedalLocomotionControllers/TemplateHelpers.h
```
Why not in `include/BipedalLocomotionControllers/GenericContainer/ *`? The files will be installed in 
https://github.com/dic-iit/bipedal-locomotion-controllers/blob/e055599aa3cbe6bc3f4976a53c358694a5c0970a/cmake/AddBipedalLocomotionControllersComponent.cmake#L87

S-Dafarra replied:
I did not notice, I'll fix it.

S-Dafarra replied:
Done in https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29/commits/2fd26da1dcaac5814f255c8e7d2ba2cea0dcae15

GiulioRomualdi commented on src/GenericContainer/include/BipedalLocomotionControllers/GenericContainer.h:
```diff
@@ -0,0 +1,289 @@
+namespace BipedalLocomotionControllers {
+    template <typename T>
+    class GenericContainer;
+}
```
Probably there is a namespace missing. 

```cpp
namespace BipedalLocomotionControllers 
{
namespace GenericContainer 
{
    template <typename T>
    class GenericContainer;
}
}
```

Too verbose right?

GiulioRomualdi replied:
in C++17 we can also use
```cpp
namespace BipedalLocomotionControllers::GenericContainer
{
    template <typename T>
    class GenericContainer;
}
```

S-Dafarra replied:
To be completely honest, I am not a huge fan of duplicated names. In general, I think we may be a little verbose. Without a ``using`` we would have
```c++
BipedalLocomotionController::GenericContainer::GenericContainer c = BipedalLocomotionController::GenericContainer::makeGenericContainer(v);
```
but on the other hand, I like if we have symmetry among components and included file (i.e. the namespace follows the inclusion path). We could use another name, maybe dropping "Generic". What would you prefer?

GiulioRomualdi replied:
Something like?
```cpp
namespace BipedalLocomotionControllers::GenericContainer
{
    template <typename T>
    class Container;
}
```

traversaro replied:
Nitpick: Container is quite a generic name, and  in C++ it has a precise meaning, see https://en.cppreference.com/w/cpp/container .  This class seems to be something that is meant just to wrap vectors that can (or cannot) be resized. For this reason I think  that a name that recall vector or array would be more clear (something like VectorWrapper or GenericVector). However, you already put a lot of effort on this, so if it permits you to go faster feel free to proceed with "GenericContainer".

S-Dafarra replied:
The hope at the beginning was that it could have been used also with ``deque``, but I did not realize that it does not have a ``data`` method, hence it cannot be used with ``span``. On hindsight, this makes sense since data in ``deque`` is not contiguous. 
Anyways, I was already planning to change the name. Given the three comments above, I could name it ``Vector`` in the namespace ``GenericContainer``.

S-Dafarra replied:
I have renamed it as
```cpp
namespace BipedalLocomotionControllers::GenericContainer
{
    template <typename T>
    class Vector;
}
```

GiulioRomualdi commented on src/GenericContainer/include/BipedalLocomotionControllers/GenericContainer.h:
```diff
@@ -0,0 +1,289 @@
+    {
+        m_span = m_resizeLambda(newSize);
+        return m_span.size() == newSize;
+    }
```
Just a small comment In general, `resize()` doesn't return a `boolean`.
I completely understand the reason why you add the `bool`. I wonder if we should also change the logic in `YarpUtilities`
https://github.com/dic-iit/bipedal-locomotion-controllers/blob/e055599aa3cbe6bc3f4976a53c358694a5c0970a/src/YarpUtilities/include/BipedalLocomotionControllers/YarpUtilities/Helper.tpp#L134-L135

Probably `YarpUtilities` may support only `GenericContainer`

S-Dafarra replied:
This is something I did not think of. I may use a different name for ``resize`` (like ``resizeContainer``), or I may add another case where I check first if the input is a ``GenericContainer`` and act accordingly, or both. What would you prefer?

GiulioRomualdi replied:
Probably the easiest way is to change the `YarpUtilities` component
Furthermore, `getVectorFromSearchable` wants as input a reference to `T`
```cpp
template <typename T>
bool getVectorFromSearchable(const yarp::os::Searchable& config, const std::string& key, T& vector);
```
Since we are moving towards the use of  `GenericContainer` probably we can change `YarpUtilities` as
```cpp
template <typename T>
bool getVectorFromSearchable(const yarp::os::Searchable& config, const std::string& key, GenericContainer<T> vector);

bool getVectorFromSearchable(const yarp::os::Searchable& config, const std::string& key, std::vector<bool>& vector);
```

Adding a **HUGE** disclaimer in the documentation saying that `GenericContainer<T>`  is passed by copy because it contains a `span` object.

S-Dafarra replied:
Notice that I explicitly deleted the copy operator. I wanted to avoid ambiguity on the copy: is the span copied or the content?
I then decided that the span is copied only in the move constructor, while the assignment operator copies the content into the pointed container. In order to copy the content, I need to make sure that I am pointing to existing containers. Clearly, in the copy constructor, this is not possible. 
We are left to decide if we want the copy constructor to copy the ``span``. I would prefer to leave it deleted to avoid confusion. What do you think?

S-Dafarra replied:
I dealt with https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29#discussion_r395166265 in https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29/commits/6bea0ccfe36aa1aa1005ebf893b2b4815d6f956e

GiulioRomualdi commented on src/GenericContainer/include/BipedalLocomotionControllers/GenericContainer.h:
```diff
@@ -0,0 +1,289 @@
+
+namespace BipedalLocomotionControllers {
+
+    enum class GenericContainerMode
```
According to the indentation style of the repo, the content of a `namespace` should not be indented 

You can load the `clang-format` file in your editor to be compliant with the style. For QTCreator you can use [this guide](https://www.vikingsoftware.com/using-clang-format-with-qtcreator/)

S-Dafarra replied:
That's me fighting against it :grin:. I was used to having an additional indentation level and I added it manually. I will remove it!

S-Dafarra replied:
I removed the indentation in https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29/commits/2fd26da1dcaac5814f255c8e7d2ba2cea0dcae15

GiulioRomualdi commented:
I tried to use the `GenericContainer` in the `IParameterHandler` and I get an error
```
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp: In function ‘void ____C_A_T_C_H____T_E_S_T____0()’:
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp:39:95: error: no matching function for call to ‘BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const char [18], BipedalLocomotionControllers::GenericContainer<int>)’
     parameterHandler->setParameter("Fibonacci Numbers", makeGenericContainer(fibonacciNumbers));
                                                                                               ^
In file included from /home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp:21:0:
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:118:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const int&)
     virtual void setParameter(const std::string& parameterName, const int& parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:118:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const int&’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:125:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const double&)
     virtual void setParameter(const std::string& parameterName, const double& parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:125:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const double&’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:132:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const string&)
     virtual void setParameter(const std::string& parameterName, const std::string& parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:132:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const string& {aka const std::__cxx11::basic_string<char>&}’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:141:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const char*)
     virtual void setParameter(const std::string& parameterName, const char* parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:141:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const char*’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:148:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const bool&)
     virtual void setParameter(const std::string& parameterName, const bool& parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:148:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const bool&’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:155:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const BipedalLocomotionControllers::GenericContainer<const int>&)
     virtual void setParameter(const std::string& parameterName, const GenericContainer<const int>& parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:155:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const BipedalLocomotionControllers::GenericContainer<const int>&’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:162:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const BipedalLocomotionControllers::GenericContainer<const double>&)
     virtual void setParameter(const std::string& parameterName, const GenericContainer<const double>& parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:162:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const BipedalLocomotionControllers::GenericContainer<const double>&’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:169:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const BipedalLocomotionControllers::GenericContainer<const std::__cxx11::basic_string<char> >&)
     virtual void setParameter(const std::string& parameterName, const GenericContainer<const std::string>& parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:169:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const BipedalLocomotionControllers::GenericContainer<const std::__cxx11::basic_string<char> >&’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:176:18: note: candidate: virtual void BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const string&, const std::vector<bool>&)
     virtual void setParameter(const std::string& parameterName, const std::vector<bool>& parameter) = 0;
                  ^~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:176:18: note:   no known conversion for argument 2 from ‘BipedalLocomotionControllers::GenericContainer<int>’ to ‘const std::vector<bool>&’
src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/build.make:62: recipe for target 'src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/ParametersHandlerYarpTest.cpp.o' failed
make[2]: *** [src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/ParametersHandlerYarpTest.cpp.o] Error 1
CMakeFiles/Makefile2:1205: recipe for target 'src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/all' failed
make[1]: *** [src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

The line of code where is failing is:

```cpp
parameterHandler->setParameter("Fibonacci Numbers", makeGenericContainer(fibonacciNumbers));
```

This is seen by the compiler as
```
BipedalLocomotionControllers::ParametersHandler::IParametersHandler::setParameter(const char [18], BipedalLocomotionControllers::GenericContainer<int>)
```

But our implementation it's 
```cpp
void setParameter(const std::string& parameterName, const GenericContainer<const int>& parameter) 
``` 
I think that the compile it's failing because of the `<const int>` template variable.  In case of `span ` `<const int>` tell you that content of the `span` cannot change.

GiulioRomualdi commented:
Removing `const` from `<const int>` solves the problem 
```cpp
void setParameter(const std::string& parameterName, const GenericContainer<int>& parameter) 
```

Probably we should understand how to propagate `const` in the template argument

GiulioRomualdi commented:
This solve the problem.
```cpp
 parameterHandler->setParameter("Fibonacci Numbers", makeGenericContainer<const std::vector<int>>(fibonacciNumbers));
```
For normal `span` the const template is automatically deduced.

S-Dafarra commented:
To be honest, I am a bit lost on the use of ``const`` and I realized that I was not very careful about its use in the methods. I would like to have a behavior closer to the one of ``std::vector`` rather then ``span``. Any suggestion?

GiulioRomualdi commented:
> would work with
> ``` 
> parameterHandler->setParameter("Fibonacci Numbers", makeGenericContainer<const std::vector<const int>>(fibonacciNumbers));
> ```
> ?

No that's the error:
```
In file included from /usr/include/x86_64-linux-gnu/c++/7/bits/c++allocator.h:33:0,
                 from /usr/include/c++/7/bits/allocator.h:46,
                 from /usr/include/c++/7/memory:63,
                 from /home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp:9:
/usr/include/c++/7/ext/new_allocator.h: In instantiation of ‘class __gnu_cxx::new_allocator<const int>’:
/usr/include/c++/7/bits/allocator.h:108:11:   required from ‘class std::allocator<const int>’
/usr/include/c++/7/bits/stl_vector.h:81:14:   required from ‘struct std::_Vector_base<const int, std::allocator<const int> >::_Vector_impl’
/usr/include/c++/7/bits/stl_vector.h:166:20:   required from ‘struct std::_Vector_base<const int, std::allocator<const int> >’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/GenericContainer/include/BipedalLocomotionControllers/TemplateHelpers.h:115:46:   recursively required by substitution of ‘template<class T> struct BipedalLocomotionControllers::has_type_member<T, std::void_t<typename T::value_type> > [with T = const std::vector<const int>]’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/GenericContainer/include/BipedalLocomotionControllers/TemplateHelpers.h:115:46:   required from ‘struct BipedalLocomotionControllers::container_data<const std::vector<const int> >’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/GenericContainer/include/BipedalLocomotionControllers/GenericContainer.h:245:60:   required by substitution of ‘template<class Class, class> BipedalLocomotionControllers::GenericContainer<typename BipedalLocomotionControllers::container_data<T>::type> BipedalLocomotionControllers::makeGenericContainer(Class&, BipedalLocomotionControllers::GenericContainerMode) [with Class = const std::vector<const int>; <template-parameter-1-2> = <missing>]’
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp:39:124:   required from here
/usr/include/c++/7/ext/new_allocator.h:93:7: error: ‘const _Tp* __gnu_cxx::new_allocator<_Tp>::address(__gnu_cxx::new_allocator<_Tp>::const_reference) const [with _Tp = const int; __gnu_cxx::new_allocator<_Tp>::const_pointer = const int*; __gnu_cxx::new_allocator<_Tp>::const_reference = const int&]’ cannot be overloaded
       address(const_reference __x) const _GLIBCXX_NOEXCEPT
       ^~~~~~~
/usr/include/c++/7/ext/new_allocator.h:89:7: error: with ‘_Tp* __gnu_cxx::new_allocator<_Tp>::address(__gnu_cxx::new_allocator<_Tp>::reference) const [with _Tp = const int; __gnu_cxx::new_allocator<_Tp>::pointer = const int*; __gnu_cxx::new_allocator<_Tp>::reference = const int&]’
       address(reference __x) const _GLIBCXX_NOEXCEPT
       ^~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp: In function ‘void ____C_A_T_C_H____T_E_S_T____0()’:
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp:39:124: error: no matching function for call to ‘makeGenericContainer<const std::vector<const int, std::allocator<const int> > >(std::vector<int>&)’
 Parameter("Fibonacci Numbers", makeGenericContainer<const std::vector<const int>>(fibonacciNumbers));
                                                                                                   ^
In file included from /home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/include/BipedalLocomotionControllers/ParametersHandler/IParametersHandler.h:17:0,
                 from /home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/ParametersHandler/tests/ParametersHandlerYarpTest.cpp:21:
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/GenericContainer/include/BipedalLocomotionControllers/GenericContainer.h:245:60: note: candidate: template<class Class, class> BipedalLocomotionControllers::GenericContainer<typename BipedalLocomotionControllers::container_data<T>::type> BipedalLocomotionControllers::makeGenericContainer(Class&, BipedalLocomotionControllers::GenericContainerMode)
     GenericContainer<typename container_data<Class>::type> makeGenericContainer(Class& input,
                                                            ^~~~~~~~~~~~~~~~~~~~
/home/gromualdi/robot-code/bipedal-locomotion-controllers_estimators/src/GenericContainer/include/BipedalLocomotionControllers/GenericContainer.h:245:60: note:   substitution of deduced template arguments resulted in errors seen above
src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/build.make:62: recipe for target 'src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/ParametersHandlerYarpTest.cpp.o' failed
make[2]: *** [src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/ParametersHandlerYarpTest.cpp.o] Error 1
CMakeFiles/Makefile2:1157: recipe for target 'src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/all' failed
make[1]: *** [src/ParametersHandler/tests/CMakeFiles/ParametersHandlerYarpUnitTests.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
```

GiulioRomualdi commented:
The template deduction in `setParameter` function i.e.  (if `fibonacciNumbers` is a `const std::vector<>`) 
```cpp 
parameterHandler->setParameter("Fibonacci Numbers", makeGenericContainer(fibonacciNumbers));
```
can be achieved by adding another version of `makeGenericContainer`. This version will create a container to `const` element .

```cpp
template <typename Class, typename = const typename container_data<Class>::type>
GenericContainer<const typename container_data<Class>::type>
makeGenericContainer(const Class& input, GenericContainerMode mode = GenericContainerMode::Fixed)
{
    static_assert(is_span_constructible<Class>::value
                      || (is_data_available<Class>::value && is_size_available<Class>::value),
                  "Cannot create a span given the provided class.");

    using value_type = typename container_data<decltype(input)>::type;

    iDynTree::Span<value_type> span;

    if constexpr (is_span_constructible<Class>::value)
    {
        span = iDynTree::make_span(input);
    } else
    {
        span = iDynTree::make_span(input.data(), input.size());
    }

    if constexpr (is_resizable<Class>::value)
    {
        if (mode == GenericContainerMode::Resizable)
        {
            std::cerr << "[GenericContainer] " << type_name<Class>()
                      << " is const and cannot be resizable. Returning a non-resizable container." << std::endl;
        }
        return GenericContainer(span);
    } else
    {
        if (mode == GenericContainerMode::Resizable)
        {
            std::cerr << "[GenericContainer] " << type_name<Class>()
                      << " is not resizable. Returning a non-resizable container." << std::endl;
        }

        return GenericContainer(span);
    }
}
```

Notice that `using value_type = typename container_data<Class>::type;` has been substitute with 
 `using value_type = typename container_data<decltype(input)>::type;` in order to propagate the `const` identifier. `using value_type = typename container_data<const Class>::type` can be also used.

Thanks to these two implementations of  `makeGenericContainer` (the `const` and the `non-const`) versions), handling `const` vectors will be easier than now. 

What do you think @S-Dafarra ?

S-Dafarra commented:
> Thanks to these two implementations of makeGenericContainer (the const and the non-const) versions), handling const vectors will be easier than now.

> What do you think @S-Dafarra ?

I have added the version you proposed. I only removed the possibility of resizing. I also dealt a little better with ``const`` and not-``const`` methods. The idea is to treat the pointed container as if it was own by the class. Basically, if ``GenericContainer`` is ``const``, the data cannot be changed, even if it was possible since the internal state does not change.

GiulioRomualdi commented on src/GenericContainer/include/BipedalLocomotionControllers/GenericContainer/Vector.h:
```diff
@@ -0,0 +1,354 @@
+    const_iterator cend() const
+    {
+        return m_span.cend();
+    }
```
I would add also the `const` version of `begin` and `end`
i.e. 
```cpp
    const_iterator begin() const
    {
        return m_span.cbegin();
    }
    const_iterator end() const
    {
        return m_span.cend();
    }
```

This will allow using Range-based for loop for the container.

Similar issue: https://github.com/robotology/yarp/issues/2179 and associated PR https://github.com/robotology/yarp/pull/2182

S-Dafarra replied:
Added in https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29/commits/1a117600b0ae7245a88ab716c719101fa2161db5

GiulioRomualdi commented:
Just a side note that I think it is not possible to change.
The following code cannot compile
```cpp
std::vector<int> element;
parameterHandler->getParameter("Fibonacci Numbers", make_vector(element, VectorResizeMode::Resizable)));
```
where `getParameter` signature is
```cpp
bool getParameter(const std::string& parameterName, const GenericContainer::Vector<int>& parameter) const;
```
since `parameter` is passed as `const &` the `resizeVector()` function is not define. The function is not `const` and cannot be transformed in `const` because when the content of the `iDynTree::Span`  is resized a new `span` is created and it is store as attribute of the `GenericContainer::Vector`.

Passing the `GenericContainer::Vector` by copy didn't solve the problem because the _copy assignment operator_ has been on purpose deleted as explained by @S-Dafarra in https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29#discussion_r395274122

Changing the `getParameter` function signature as: 
```cpp
bool getParameter(const std::string& parameterName, GenericContainer::Vector<int>& parameter) const;
```
solved the problem. The drawback is that the `parameter` must be an _l-value_ and another line haas to be added to the code.
```cpp
std::vector<int> element;
auto container = make_vector(element, VectorResizeMode::Resizable); // <---- container is an l-value
parameterHandler->getParameter("Fibonacci Numbers", container)); // <---- the contect of container (i.e. element) is resized
```

S-Dafarra commented:
To be honest I kind of like the version
```
bool getParameter(const std::string& parameterName, GenericContainer::Vector<int>& parameter) const;
```
since it makes it clear that that method is going to modify "parameter". What if we create a new function ``make_vector_ptr`` which returns a ``shared_ptr<Vector>``. Then you can pass this by copy. Can you try if this would work?

S-Dafarra commented:
I am little puzzled by this compile error on Windows:
```
GenericContainerTest.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const Catch::Detail::unprintableString" (?unprintableString@Detail@Catch@@3V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@B) referenced in function "private: virtual void __cdecl Catch::BinaryExpr<class std::reverse_iterator<class iDynTree::details::span_iterator<class iDynTree::Span<int,-1>,0> > const &,class std::reverse_iterator<class iDynTree::details::span_iterator<class iDynTree::Span<int,-1>,0> > const &>::streamReconstructedExpression(class std::basic_ostream<char,struct std::char_traits<char> > &)const " (?streamReconstructedExpression@?$BinaryExpr@AEBV?$reverse_iterator@V?$span_iterator@V?$Span@H$0?0@iDynTree@@$0A@@details@iDynTree@@@std@@AEBV12@@Catch@@EEBAXAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@@Z) [D:\a\bipedal-locomotion-controllers\bipedal-locomotion-controllers\build\src\GenericContainer\tests\GenericContainerUnitTests.vcxproj]
```

GiulioRomualdi commented:
> I am little puzzled by this compile error on Windows:
> ```
> GenericContainerTest.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const Catch::Detail::unprintableString" (?unprintableString@Detail@Catch@@3V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@B) referenced in function "private: virtual void __cdecl Catch::BinaryExpr<class std::reverse_iterator<class iDynTree::details::span_iterator<class iDynTree::Span<int,-1>,0> > const &,class std::reverse_iterator<class iDynTree::details::span_iterator<class iDynTree::Span<int,-1>,0> > const &>::streamReconstructedExpression(class std::basic_ostream<char,struct std::char_traits<char> > &)const " (?streamReconstructedExpression@?$BinaryExpr@AEBV?$reverse_iterator@V?$span_iterator@V?$Span@H$0?0@iDynTree@@$0A@@details@iDynTree@@@std@@AEBV12@@Catch@@EEBAXAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@@Z) [D:\a\bipedal-locomotion-controllers\bipedal-locomotion-controllers\build\src\GenericContainer\tests\GenericContainerUnitTests.vcxproj]
> ```
> 
> Any idea? 

I can try to compile on my windows machine (yarp and iDynTree should be updated)

GiulioRomualdi commented:
When I tried to configure the project on windows I got the following error
```
Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.17134.
The C compiler identification is MSVC 19.24.28315.0
The CXX compiler identification is MSVC 19.24.28315.0
Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.24.28314/bin/Hostx64/x64/cl.exe
Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.24.28314/bin/Hostx64/x64/cl.exe -- works
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Detecting C compile features
Detecting C compile features - done
Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.24.28314/bin/Hostx64/x64/cl.exe
Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.24.28314/bin/Hostx64/x64/cl.exe -- works
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Detecting CXX compile features
Detecting CXX compile features - done
Found YARP: C:/Users/gromualdi/robot-code/robotology-superbuild/build/install/lib/cmake/YARP (found version "3.3.1+3-20200117.4+git5a112f9c9")
Using iCub from install
Found YARP: C:/Users/gromualdi/robot-code/robotology-superbuild/build/install/lib/cmake/YARP (found version "3.3.1+3-20200117.4+git5a112f9c9")
CMake Warning (dev) at C:/Users/gromualdi/robot-code/vcpkg/scripts/buildsystems/vcpkg.cmake:285 (_find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable Eigen3_ROOT is set to:

    C:\Program Files\robotology\eigen-3.3.4

  For compatibility, CMake is ignoring the variable.
Call Stack (most recent call first):
  C:/Users/gromualdi/robot-code/robotology-superbuild/build/install/CMake/OsqpEigenConfig.cmake:6 (find_package)
  C:/Users/gromualdi/robot-code/vcpkg/scripts/buildsystems/vcpkg.cmake:285 (_find_package)
  C:/Program Files/CMake/share/cmake-3.16/Modules/CMakeFindDependencyMacro.cmake:47 (find_package)
  C:/Users/gromualdi/robot-code/robotology-superbuild/build/install/lib/cmake/iDynTree/iDynTreeConfig.cmake:24 (find_dependency)
  C:/Users/gromualdi/robot-code/vcpkg/scripts/buildsystems/vcpkg.cmake:285 (_find_package)
  cmake/BipedalLocomotionControllersFindDependencies.cmake:120 (find_package)
  CMakeLists.txt:20 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

Found Eigen3: C:/Users/gromualdi/robot-code/vcpkg/installed/x64-windows/include/eigen3 (Required is at least version "2.91.0") 
Found YARP: C:/Users/gromualdi/robot-code/robotology-superbuild/build/install/lib/cmake/YARP (found version "3.3.1+3-20200117.4+git5a112f9c9")
CMake Error at CMakeLists.txt:70 (include):
  include could not find load file:

    AddBipedalLocomotionControllersUnitTest


CMake Error at CMakeLists.txt:71 (include):
  include could not find load file:

    AddBipedalLocomotionControllersComponent


CMake Error at src/CMakeLists.txt:5 (add_subdirectory):
  add_subdirectory given source "GenericContainer" which is not an existing
  directory.


CMake Error at src/YarpUtilities/CMakeLists.txt:8 (add_bipedal_component):
  Unknown CMake command "add_bipedal_component".


Configuring incomplete, errors occurred!
See also "C:/Users/gromualdi/robot-code/bipedal-locomotion-controllers/build/CMakeFiles/CMakeOutput.log".
```

S-Dafarra commented:
It seems that you updated only the ``CMakelists``. That's weird

GiulioRomualdi commented:
> It seems that you updated only the CMakelists. That's weird

yes my fault. I'm getting your same error https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29#issuecomment-601782582. I'm investigating on it

S-Dafarra commented:
> yes my fault. I'm getting your same error #29 (comment). I'm investigating on it

That's a good news:grin: I tried to take out from ``REQUIRE()`` all the expressions involving an iterator. I can't really understand why there is this missing symbol. A sort of missing conversion from ``iterator`` to ``string`` inside the ``Catch`` machinery?

GiulioRomualdi commented:
Changing the test as 
```cpp
    SECTION("iterator_default_init")
    {
        GenericContainer::Vector<int>::iterator it1;
        GenericContainer::Vector<int>::iterator it2;
        bool ok = it1 == it2;
        REQUIRE(ok);
    }
```
makes the code compiling on Windows visualstudio 2019. Something fishy is happening in the CATCH2 macro. Shall we open an issue there?

GiulioRomualdi commented:
Bingo! Adding a parenthesis solves the problem. 
```cpp
REQUIRE((it1 == it2)); // < ---- compile
REQUIRE(it1 == it2); // < ---- does not compile
```
Since the macro is something strictly depended on the compiler probably this problem happens only on Windows

S-Dafarra commented:
> Bingo! Adding a parenthesis solves the problem.

Thanks a lot @GiulioRomualdi! In https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29/commits/02eb1c3ab47360aa44d32e42d7ab92bec4ad4744 I implemented what proposed in this comment: https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29#issuecomment-601748866

GiulioRomualdi commented:
>  I implemented what proposed in this comment: https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29#issuecomment-601748866

I don't know if it is a good idea to change the interface of the parameters handler adding the `shared_ptr` as input. i.e. 
```cpp
bool getParameter(const std::string& parameterName, std::shared_ptr<GenericContainer::Vector<int>> parameter) const
```
Since it will create the asimmetry with the other `getParameter` functions, i.e.  
```cpp
bool getParameter(const std::string& parameterName, std::vector<bool>& parameter) const
bool getParameter(const std::string& parameterName, int& parameter) const
```

Furthermore, it may over complexify the `YarpImplementation` of the `ParameterHandler` class. 
For instance, it will not work anymore. 
https://github.com/dic-iit/bipedal-locomotion-controllers/blob/21d5ed943bfbe7b364896ede6487ea634800cd11/src/ParametersHandler/YarpImplementation/include/BipedalLocomotionControllers/ParametersHandler/YarpImplementation.tpp#L18-L33

and a specific function for each datatype has to be implemented

S-Dafarra commented:
Ah too bad. Yes, I agree it is better to keep as much as possible the interface clean. Maybe I can write a template specialization for the vector of bool. Do you have any idea to avoid that additional line? Should we enable the copy constructor?

GiulioRomualdi commented:
I think it's fine like this. This is how the parameters handler implementation is now. https://github.com/GiulioRomualdi/bipedal-locomotion-controllers/tree/feature/improve_parameter_handler_without_shared_ptr

If you more or less agree with the implementation, I think you can add some lines of documentation and the we merge the PR

S-Dafarra commented:
> If you more or less agree with the implementation, I think you can add some lines of documentation and the we merge the PR

Deal!

What if we add a templated method in ``IParametersHandler`` like
```cpp
template<class Vector, typename = someCheck>
bool getParameter(std::string& parameterName, Vector& parameter,
                               VectorResizeMode mode = VectorResizeMode::Fixed)
{
    auto genericVector = make_vector(parameter, mode);
    return getParameter(parameterName, genericVector);
}

```
where ``someCheck`` checks if ``Vector``is a suitable class and is not a ``GenericContainer::Vector`` to avoid ambiguity (and infinite recursive loop). Is it worth it?

GiulioRomualdi commented:
> > If you more or less agree with the implementation, I think you can add some lines of documentation and the we merge the PR
> 
> Deal!
> 
> What if we add a templated method in `IParametersHandler` like
> 
> ```c++
> template<class Vector, typename = someCheck>
> bool getParameter(std::string& parameterName, Vector& parameter,
>                                VectorResizeMode mode = VectorResizeMode::Fixed)
> {
>     auto genericVector = make_vector(parameter, mode);
>     return getParameter(parameterName, genericVector);
> }
> ```
> 
> where `someCheck` checks if `Vector`is a suitable class and is not a `GenericContainer::Vector` to avoid ambiguity (and infinite recursive loop). Is it worth it?

I think it's ok. For this, we can move the discussion in #28

S-Dafarra commented:
> where ``someCheck`` checks if ``Vector`` is a suitable class and is not a ``GenericContainer::Vector`` to avoid ambiguity (and infinite recursive loop).

After https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29/commits/9c2764f101d8b64832ef993ebd9e0e633f1b2651, ``someCheck`` should be substituted by
```c++
std::enable_if<!is_vector<Vector>::value && !std::is_same<Vector, std::string>::value && 
                        is_vector_constructible<Vector>::value>::type>
```
i.e. ``Vector`` should not be a ``GenericContainer::Vector``, neither a ``std::string`` and should be a supported type for ``GenericContainer::Vector``.

With https://github.com/dic-iit/bipedal-locomotion-controllers/pull/29/commits/ebba18a76824a4126cbdee5c4ec65e098438f4d3 I added documentation and modified the Changelog.

GiulioRomualdi commented:
If you agree we can merge!

S-Dafarra commented:
Yes, I agree
